src/connector/src/sink/encoder/json.rs

```diff
@@ -262,6 +283,95 @@ fn datum_to_json_object(
     Ok(value)
 }
 
+fn json_converter_with_schema<'a>(
+    object: Value,
+    name: String,
+    fields: impl Iterator<Item = &'a Field>,
+) -> Map<String, Value> {
+    let mut mapping = Map::with_capacity(2);
+    mapping.insert(
+        "schema".to_string(),
+        json!({
+            "type": "struct",
+            "fields": fields.map(|field| {
+                let mut mapping = type_as_json_schema(&field.data_type);
+                mapping.insert("field".to_string(), json!(field.name));
+                mapping
+            }).collect_vec(),
+            "optional": "false",
+            "name": name,
+        }),
+    );
+    mapping.insert("payload".to_string(), object);
+    mapping
+}
+
+// reference: https://github.com/apache/kafka/blob/80982c4ae3fe6be127b48ec09caff11ab5f87c69/connect/json/src/main/java/org/apache/kafka/connect/json/JsonSchema.java#L39
+pub(crate) fn schema_type_mapping(rw_type: &DataType) -> &'static str {
+    match rw_type {
+        DataType::Boolean => "boolean",
+        DataType::Int16 => "int16",
+        DataType::Int32 => "int32",
+        DataType::Int64 => "int64",
+        DataType::Float32 => "float",
+        DataType::Float64 => "double",
+        DataType::Decimal => "string",
+        DataType::Date => "int32",
+        DataType::Varchar => "string",
+        DataType::Time => "int64",
+        DataType::Timestamp => "int64",
+        DataType::Timestamptz => "string",
+        DataType::Interval => "string",
+        DataType::Struct(_) => "struct",
+        DataType::List(_) => "array",
+        DataType::Bytea => "bytes",
+        DataType::Jsonb => "string",
+        DataType::Serial => "int32",
+        DataType::Int256 => "string",
+    }
+}
+
+fn type_as_json_schema(rw_type: &DataType) -> Map<String, Value> {
+    let mut mapping = Map::with_capacity(4); // type + optional + fields/items + field
+    mapping.insert("type".to_string(), json!(schema_type_mapping(rw_type)));
+    mapping.insert("optional".to_string(), json!("true"));
+    match rw_type {
+        DataType::Struct(struct_type) => {
+            let mut sub_fields = Vec::new();
+            let map_sub_struct = |(sub_name, sub_type): (String, &DataType)| {
+                let mut sub_mapping = type_as_json_schema(sub_type);
+                sub_mapping.insert("field".to_string(), json!(sub_name));
+                sub_fields.push(sub_mapping);
+            };
+            if struct_type.is_unnamed() {
+                // unnamed struct, rename with `f{idx+1}`
+                struct_type
+                    .types()
+                    .enumerate()
+                    .map(|(idx, sub_type)| (format!("f{}", idx + 1), sub_type))
+                    .for_each(map_sub_struct);
+            } else {
+                struct_type
+                    .iter()
+                    .map(|(sub_name, sub_type)| (sub_name.to_string(), sub_type))
+                    .for_each(map_sub_struct);
+            }
```

Comment:
Good catch. This logic of handling of struct with unnamed fields should be applicable to other places as well. For now let's just ignore them and use the same call as others: `struct_type.iter()`

Reply:
Do you mean that I should delete the `if struct_type.is_unnamed()` branch, and keep only:
```rust
                struct_type
                    .iter()
                    .map(|(sub_name, sub_type)| (sub_name.to_string(), sub_type))
                    .for_each(map_sub_struct);
```

Reply:
Yes. Or in a slightly better form: `map().collect()` rather than `for_each()` on `mut Vec`